<!DOCTYPE html>

<html xml:lang="en" lang="en">
<link rel='shortcut icon' type='image/x-icon' href='favicon.ico' />
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Digital Expression Explorer 2
</title>
<style type="text/css">

 body{
  -webkit-text-size-adjust:none;
  margin: 20px auto;
  max-width:720px;
  line-height:1.6;
  font:normal normal normal 100% verdana,sans-serif;
  font-size:22px;
  color:#3a3a3a;
  padding:0 10px
 }

 h1,h2,h3{
  line-height:1.2;
  color:#406fef ;
 }

 aside,p,ul{
  color:#3a3a3a;
 }

 table,th,tr,td{
   line-height:1.2;
   font-family:sans-serif;
   font-size:20px;
   background-color:transparent;
   padding: 2px;
  }

ul {
    list-style-type: none;
    margin: 0ex;
    padding: 0;
}

li {
    display: inline;
}

.button {
    background-color: #999999;
    color: white;
    padding: 10px 25px;
    text-decoration: none;
#    font-size: 1.8vmax;
    cursor: pointer;
    -webkit-transition-duration: 0.4s; /* Safari */
    transition-duration: 0.4s;
}

.button:hover {
    background-color: #406fef; /* blue */
    color: white;
}

 input{
  font-size:2vmax;
 }

/* The container */
.container {
    display: block;
    position: relative;
    padding-left: 35px;
    margin-bottom: 12px;
    cursor: pointer;
    font-size: 22px;
    color:#3a3a3a;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

/* Hide the browser's default radio button */
.container input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
}

/* Create a custom radio button */
.checkmark {
    position: absolute;
    top: 0;
    left: 0;
    height: 25px;
    width: 25px;
    background-color: #eee;
    border-radius: 50%;
}

/* On mouse-over, add a grey background color */
.container:hover input ~ .checkmark {
    background-color: #ccc;
}

/* When the radio button is checked, add a blue background */
.container input:checked ~ .checkmark {
    background-color: #406fef;
}

/* Create the indicator (the dot/circle - hidden when not checked) */
.checkmark:after {
    content: "";
    position: absolute;
    display: none;
}

/* Show the indicator (dot/circle) when checked */
.container input:checked ~ .checkmark:after {
    display: block;
}

/* Style the indicator (dot/circle) */
.container .checkmark:after {
 	top: 9px;
	left: 9px;
	width: 8px;
	height: 8px;
	border-radius: 50%;
	background: white;
}

	</style>
</head>
<body>

<ul>
<a href="/index.html" class="button">HOME</a>
<a href="/about.html" class="button">ABOUT</a>
<a href="/help.html" class="button">HELP</a>
<a href="/bulk.html" class="button">BULK DATA</a>
</ul>


<header>
<h1 id="content">About DEE2</h1>

<p>The compendium is designed to bring biologists closer to large scale gene expression data sets.
We have processed thousands of public RNA-seq data sets from a veriety of organisms with open-source bioinformatics tools and make them 
freely accessible.
</p>

<hr>
<h2 id="content">About us</h2>

<p>The compendium is brought to you by Mark Ziemann (Deakin University) with support from the <a href="https://www.monash.edu/medicine/ccs/diabetes/research/el-osta-group" target="_blank">
Epigenetics in Human Health and Disease Laboratory</a>  and 
<a href="https://www.monash.edu/researchinfrastructure/eresearch" target="_blank">
Monash eResearch Centre.</a>
 We value your feedback, so feel free to contact us by email (mark.ziemann[at]gmail.com) or raise an 
issue on our <a href="https://github.com/markziemann/dee2/" target="_blank">  GitHub Repo</a>.
</p>

<hr>
<h2 id="content">Acknowledgements</h2>
<p>A project of this size cannot be done by a single person or team, so we acknowledge support from the following:
<br><br>
<a href="https://nectar.org.au" target="_blank">Nectar Research Cloud</a>, a collaborative 
Australian research platform supported by the National Collaborative Research Infrastructure Strategy (NCRIS). </li>
<br><br>
The Multi-modal Australian ScienceS Imaging and Visualisation Environment 
(<a href="http://www.massive.org.au" target="_blank">MASSIVE</a>).
<br><br>
<a href="http://www.deakin.edu.au/research/support-for-researchers/eresearch", target="_blank">Deakin eResearch</a>, 
<a href="https://www.monash.edu/researchinfrastructure/eresearch", target="_blank">Monash eResearch Centre</a> 
and the IT department of <a href="https://www.baker.edu.au/", target="_blank">Baker Heart and Diabetes Institute</a>.
<br><br>
This research would not have been possible without facilities and help from NCBI. In particular, we acknowledge 
support from <a href="http://www.ncbi.nlm.nih.gov/sra"> SRA</a> and 
<a href="http://www.ncbi.nlm.nih.gov/geo"> GEO</a> for curating and hosting these data.
</p>

<hr>
<hr>

<h2 id="content">Latest News</h2>

<p> 28th November 2018 - DEE2 was presented at the ABACBS2018 conference in Melbourne. See the <a href="https://doi.org/10.5281/zenodo.1561839">Poster</a>.
</p>

<p> 28th November 2018 - A guide to loading in bulk dumps had been added to my <a href="https://genomespot.blogspot.com/2018/11/using-dee2-bulk-data-dumps.html">Blog</a>. 
</p>

<p> 12th November 2018 - Ever wondered how accurate DEE2 data actually is? Well we have just undertaken a simulation 
study and comparison to GEO deposited data to demonstrate the quality of DEE2 data. More details 
<a href="https://github.com/markziemann/dee2/tree/master/validation"> here</a>.
</p>

<p> 26th October 2018 - Recently there have been many projects with &#8805;500 SRA runs, meaning that they have not 
been readily available from the webserver as one file. In order to address this, I've packaged these SRA projects with 
&#8805;200 SRA run numbers into zip files. Only projects with EVERY run processed successfully by DEE2 are included.
These are now available <a href="/huge">here</a>.
</p>

<p>7th October 2018 - QC information is now linked in the search results and available by hovering the 
mouse over the QC link. Note that all the datasets are currently classified as "PASS" but this will be
chenging soon. Also we moved to another nectar server with storage so there have been some updates to 
webserver cgi scripts. Also the keyword search is now case insensitive.
</p>

<a href="/news.html">More news</a>

<hr>
<h2 id="content">Data processing</h2>

<p>Our data processing procedure entails:
</p>

<ol>
<li> -Download from NCBI SRA</li><br>
<li> -Diagnose sequence format</li><br>
<li> -Sequence quality trimming and adapter clipping</li><br>
<li> -Alignment to genome and transcriptome</li><br>
<li> -Assignment of reads to genes and transcripts</li>
</ol>

<p>More information regarding the data processing method is available at the  <a href="https://github.com/markziemann/dee2" target="_blank"> GitHub repo</a>. 
Below are the versions and major parameters used in the pipeline.
</p>

<style type="text/css">
.tg1  {border-collapse:collapse;border-spacing:0;}
.tg1 td{font-family:Arial, sans-serif;font-size:12px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg1 th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg1 .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg1">
  <tr>
    <th class="tg-0lax" colspan="4">Software versions and parameters used in the pipeline.</th>
  </tr>
  <tr>
    <td class="tg-0lax" rowspan="2"><span style="font-weight:bold">Software, version</span></td>
    <td class="tg-0lax" rowspan="2"><span style="font-weight:700">Purpose</span></td>
    <td class="tg-0lax" colspan="2"><span style="font-weight:bold">Parameter</span></td>
  </tr>
  <tr>
    <td class="tg-0lax"><span style="font-weight:bold">SE</span></td>
    <td class="tg-0lax"><span style="font-weight:bold">PE</span></td>
  </tr>
  <tr>
    <td class="tg-0lax">Aspera client, v3.5.4</td>
    <td class="tg-0lax">Rapid download of sequence data</td>
    <td class="tg-0lax" colspan="2">ascp -l 500m -O 33001 -T -i $ID $URL .</td>
  </tr>
  <tr>
    <td class="tg-0lax" rowspan="3">SRA toolkit, v2.8.2</td>
    <td class="tg-0lax">Validate downloaded SRA files</td>
    <td class="tg-0lax" colspan="2">vdb-validate $SRA</td>
  </tr>
  <tr>
    <td class="tg-0lax">diagnose single or paired end</td>
    <td class="tg-0lax" colspan="2">fastq-dump -X 4000 --split-files $SRA</td>
  </tr>
  <tr>
    <td class="tg-0lax">dump fastq</td>
    <td class="tg-0lax" colspan="2">(see parallel-fastq-dump below)</td>
  </tr>
  <tr>
    <td class="tg-0lax">FastQC, v0.11.5</td>
    <td class="tg-0lax">Diagnose basespace / colorspace, quality encoding, read length from 4000 reads</td>
    <td class="tg-0lax">fastqc $FQ1</td>
    <td class="tg-0lax">fastqc $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">parallel-fastq-dump, 0.6.3</td>
    <td class="tg-0lax">Rapid decompression of sequence data from .sra files</td>
    <td class="tg-0lax" colspan="2">parallel-fastq-dump --threads $THREADS --outdir . --split-files --defline-qual + -s ${SRR}.sra</td>
  </tr>
  <tr>
    <td class="tg-0lax" rowspan="3">Skewer, v0.2.2</td>
    <td class="tg-0lax">3’ quality trimming</td>
    <td class="tg-0lax">skewer -l 18 -q 10 -k inf -t $THREADS -o $SRR $FQ1</td>
    <td class="tg-0lax">skewer -l 18 -q 10 -k inf -t $THREADS -o $SRR $FQ1 $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">Adapter clipping</td>
    <td class="tg-0lax">skewer -l 18 -t $THREADS -x $ADAPTER -o $SRR $FQ1</td>
    <td class="tg-0lax">skewer -l 18 -t $THREADS -x $ADAPTER1 -y $ADAPTER2 -o $SRR $FQ1 $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">5’ trimming</td>
    <td class="tg-0lax">skewer -m ap --cut $CLIP_NUM,$CLIP_NUM -l 18 -k inf -t $THREADS $FQ1</td>
    <td class="tg-0lax">skewer -m ap --cut $R1_CLIP_NUM,$R2_CLIP_NUM -l 18 -k inf -t $THREADS $FQ1 $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">Minion, v13-100</td>
    <td class="tg-0lax">3’ adapter detection</td>
    <td class="tg-0lax">minion search-adapter -i $FQ1</td>
    <td class="tg-0lax">minion search-adapter -i $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">Bowtie2, v2.3.2</td>
    <td class="tg-0lax">Adapter contamination detection</td>
    <td class="tg-0lax" colspan="2">bowtie2 -f -x $BT2_REF -S /dev/stdout $ADAPTER</td>
  </tr>
  <tr>
    <td class="tg-0lax">FASTX-Toolkit, v0.0.14</td>
    <td class="tg-0lax">Progressive 5’ trimming</td>
    <td class="tg-0lax">fastx_trimmer -f {5,9,13,21} -m 18 -Q 33 -i  $FQ1</td>
    <td class="tg-0lax">fastx_trimmer -f {5,9,13,21} -m 18 -Q 33 -i  $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">STAR v020201</td>
    <td class="tg-0lax">Gene-level mapping, Diagnose strandedness</td>
    <td class="tg-0lax">STAR --runThreadN $THREADS --quantMode GeneCounts \<br>--genomeLoad LoadAndKeep  --outSAMtype None \<br>--genomeDir $STAR_DIR --readFilesIn=$FQ1</td>
    <td class="tg-0lax">STAR --runThreadN $THREADS --quantMode GeneCounts \<br>--genomeLoad LoadAndKeep --outSAMtype None \<br>--genomeDir $STAR_DIR --readFilesIn=$FQ1 $FQ2</td>
  </tr>
  <tr>
    <td class="tg-0lax">Kallisto, v0.43.1</td>
    <td class="tg-0lax">Transcript-level mapping</td>
    <td class="tg-0lax">kallisto quant $KALLISTO_STRAND_PARAMETER \<br>--single -l 100 -s 20 -t $THREADS -o . \<br> -i $KAL_REF $FQ1</td>
    <td class="tg-0lax">kallisto quant $KALLISTO_STRAND_PARAMETER \<br>-t $THREADS -o . -i $KAL_REF $FQ1 $FQ2</td>
  </tr>
</table>

<hr>
<hr>
<h2 id="content">Reference genome information</h2>

<p>The compendium relies on reference genome sequence and annotation information provided by 
<a href="http://ensemblgenomes.org/" target="_blank" > Ensembl Genomes </a>. 
</p>


<style type="text/css">
.tg2  {border-collapse:collapse;border-spacing:0;}
.tg2 td{font-family:Arial, sans-serif;font-size:18px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg2 th{font-family:Arial, sans-serif;font-size:18px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg2 .tg-s268{text-align:left}
</style>
<table class="tg2">
  <tr>
    <th class="tg-s268"><strong>Species</strong></th>
    <th class="tg-s268"><strong>Genome Reference Sequence and Annotation</strong>
  </th>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Arabidopsis thaliana</span></td>
    <td class="tg-s268">Ensembl release 36<br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/release-36/plants/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/release-36/plants/gtf/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.36.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/release-36/plants/fasta/arabidopsis_thaliana/cdna/Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Caenorhabditis elegans</span><br></td>
    <td class="tg-s268">Ensembl release 90<br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/fasta/caenorhabditis_elegans/cdna/Caenorhabditis_elegans.WBcel235.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Drosophila melanogaster</span></td>
    <td class="tg-s268">Ensembl release 90<br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/drosophila_melanogaster/dna/Drosophila_melanogaster.BDGP6.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/fasta/drosophila_melanogaster/cdna/Drosophila_melanogaster.BDGP6.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Danio rerio</span></td>
    <td class="tg-s268">Ensembl release 90<br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/danio_rerio/dna/Danio_rerio.GRCz10.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/danio_rerio/Danio_rerio.GRCz10.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/danio_rerio/cdna/Danio_rerio.GRCz10.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Escherichia coli</span></td>
    <td class="tg-s268">Ensembl release 36<br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-36/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/dna/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-36/gtf/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.36.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-36/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cdna/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Homo sapiens</span></td>
    <td class="tg-s268">Ensembl release 90<br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Mus musculus</span></td>
    <td class="tg-s268">Ensembl release 90<br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna_sm.primary_assembly.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/mus_musculus/Mus_musculus.GRCm38.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Rattus norvegicus</span></td>
    <td class="tg-s268">Ensembl release 90 <br>
     <a href="ftp://ftp.ensemblorg.ebi.ac.uk/pub/release-90/fasta/rattus_norvegicus/dna/Rattus_norvegicus.Rnor_6.0.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.90.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensembl.org/pub/release-90/fasta/rattus_norvegicus/cdna/Rattus_norvegicus.Rnor_6.0.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>
  </td>
  </tr>
  <tr>
    <td class="tg-s268"><span style="font-style:italic">Saccharomyces cerevisiae</span></td>
    <td class="tg-s268">Ensembl release 36<br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/fungi/release-36/fasta/saccharomyces_cerevisiae/dna/Saccharomyces_cerevisiae.R64-1-1.dna_sm.toplevel.fa.gz">Genome sequence (fasta)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/release-36/fungi/gtf/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.36.gtf.gz">Gene annotation set (GTF)</a><br>
     <a href="ftp://ftp.ensemblgenomes.org/pub/release-36/fungi/fasta/saccharomyces_cerevisiae/cdna/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz">cDNA sequences (fasta)</a><br>  
  </td>
  </tr>
</table>

<br>


<hr>
<h2 id="content">Update schedule</h2>

<p>We are updating the compendium fortnightly. Upon release of an updated genome build, we intend to update the data for that organism within a year, keeping a
previously archived version for bulk download only. Gene annotation sets will not be updated windependent of the genome build.
</p>

</div>

</body>
</html>
