<!DOCTYPE html>

<html xml:lang="en" lang="en">
<link rel='shortcut icon' type='image/x-icon' href='favicon.ico' />
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Digital Expression Explorer 2
</title>
<style type="text/css">

 body{
  -webkit-text-size-adjust:none;
  margin: 20px auto;
  max-width:760px;
  line-height:1.6;
  font:normal normal normal 100% verdana,sans-serif;
  font-size:22px;
  color:#3a3a3a;
  padding:0 10px
 }

 h1,h2,h3{
  line-height:1.2;
  color:#406fef ;
 }

 aside,p,ul{
  color:#3a3a3a;
 }

 table,th,tr,td{
   line-height:1.2;
   font-family:sans-serif;
   font-size:20px;
   background-color:transparent;
   padding: 2px;
   border: 1px solid black;
   border-collapse: collapse;
  }

ul {
    list-style-type: none;
    margin: 0ex;
    padding: 0;
}

li {
    display: inline;
}

.button {
    background-color: #999999;
    color: white;
    padding: 8px 25px;
    leading: 3;
    line-height: 2em;
    text-decoration: none;
#    font-size: 1.8vmax;
    cursor: pointer;
    -webkit-transition-duration: 0.4s; /* Safari */
    transition-duration: 0.4s;
}

.button:hover {
    background-color: #406fef; /* blue */
    color: white;
}

 input{
  #font-size:4vmax;
  #font-size:400%;

 }

/* The container */
.container {
    display: block;
    position: relative;
    padding-left: 35px;
    margin-bottom: 12px;
    cursor: pointer;
    font-size: 22px;
    color:#3a3a3a;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

/* Hide the browser's default radio button */
.container input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
}

/* Create a custom radio button */
.checkmark {
    position: absolute;
    top: 0;
    left: 0;
    height: 25px;
    width: 25px;
    background-color: #eee;
    border-radius: 50%;
}

/* On mouse-over, add a grey background color */
.container:hover input ~ .checkmark {
    background-color: #ccc;
}

/* When the radio button is checked, add a blue background */
.container input:checked ~ .checkmark {
    background-color: #406fef;
}

/* Create the indicator (the dot/circle - hidden when not checked) */
.checkmark:after {
    content: "";
    position: absolute;
    display: none;
}

/* Show the indicator (dot/circle) when checked */
.container input:checked ~ .checkmark:after {
    display: block;
}

/* Style the indicator (dot/circle) */
.container .checkmark:after {
 	top: 9px;
	left: 9px;
	width: 8px;
	height: 8px;
	border-radius: 50%;
	background: white;
}


	</style>
</head>
<body>

<ul>
<a href="/index.html" class="button">HOME</a>
<a href="/about.html" class="button">ABOUT</a>
<a href="/help.html" class="button">HELP</a>
<a href="/bulk.html" class="button">BULK DATA</a>
<a href="/genesets.html" class="button">GENE SETS</a>
</ul>

<h1 id="content">Help@DEE2</h1>

<h2 id="content">FAQ</h2>

<p>Q) My datasets of interest aren't included in DEE. What should I do?
</p>

<ol>
 <li>A) Firstly confirm that the species/accession number combination is correct. Then check whether the sequence data is 
available from the SRA ftp site. Check that the study raw data is released and not under embargo. If the dataset has been 
added recently, then you have a few options. You can run the docker or singularity image on your own server or on the cloud. 
The instructions to do this are in the github README.md <a href="https://github.com/markziemann/dee2" target="_blank" >here</a>.
Alternatively contact us and we'll have it added.</li>
</ol>
<hr>

<p>Q) How were the QC metrics generated and what do they mean?
</p>
<ol>
 <li> I have provided an explanation of each of the QC metrics on the GitHub page 
 <a href="https://github.com/markziemann/dee2/blob/master/qc/qc_metrics.md" >here</a>.
</ol>
<hr>

<p>Q) Can I get the TPM values from Kallisto?
</p>

<ol>
 <li>A) Yes, the abundance.tsv files containing the TPM values are available for download for each SRA run (eg: <a href="http://dee2.io/data/dmelanogaster/SRR1991949/SRR1991949.ke.tsv.gz" target="_blank">http://dee2.io/data/dmelanogaster/SRR1991949/SRR1991949.ke.tsv.gz</a>).
</ol>

<hr>
<p>Q) How do I collapse transcript counts to the gene level?
</p>

<ol>
 <li> You can use the getDEE2 R package documented <a href="https://github.com/markziemann/dee2/blob/master/AccessDEEfromR.md" target="_blank" >here</a>. It performs a simple aggregation by sum of isoforms to the parent gene. 
</ol>

<hr>

<p>Q) How do I open zip compressed files?
</p>

<ol>
 <li>A) Winzip can uncompress .zip files and is available <a href="http://www.sfsu.edu/ftp/win/utils/winzip80.exe" target="_blank"> 
here</a> for free. Decompression tools are also available from the Apple and Android app store for mobile devices.</li>
</ol>

<hr>

<p>Q) What tools are recommended for analyzing RNA-seq count data?
</p>

<ol>
 <li>A) There are many options, but we recommend <a href="http://degust.erc.monash.edu/" target="_blank"> Degust</a> 
for new users.</li>
</ol>
<hr>

<p>Q) How do I load the data into R?
</p>

<ol>
 <li>A) We have written an R/Bioconductor package that interfaces with the webpage and loads data into R. For more
information, read the documentation <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/getDEE2/inst/doc/getDEE2.html"  target="_blank" >here</a> </li> .
</ol>
<hr>

<p>Q) The webserver is limited to 500 datasets but my project of interest contains more than that - can I still use DEE2?
</p>

<ol>
 <li>A) Yes, you have three options - the best one is to see whether the project has been packaged. For convenience, we have packaged all 
projects with >200 runs that have been completely processed by DEE2 into zip files which contain the expression data. These are accessible 
<a href="http://dee2.io/huge/" target="_blank">here</a>. If the project is not included there, you could download the project in chunks by
submitting 500 run accessions at a time. Thirdly, you could download the full bulk data and filter as necessary.
</ol>
<hr>

<p>Q) Can I load the data directly into my Galaxy history?
</p>

<ol>
 <li>A) DEE data can be uploaded manually into Galaxy; we will be investigating direct integration in future.</li>
</ol>
<hr>

<p>Q) My DEE data contains multiple technical replicates, how can I aggregate the counts?
</p>

<ol>
 <li>A) The getDEE2 R package has a function called srx_agg() which can do this.
Refer to the <a href="https://www.bioconductor.org/packages/release/bioc/manuals/getDEE2/man/getDEE2.pdf">package manual</a> for details. 
Alternatives include <a href='http://stackoverflow.com/questions/10464055/use-awk-to-calculate-sum-for-multiple-columns'> awk </a> or even 
spreadsheet software (ie. Excel). If using Excel or other spreadsheet software, please be wary of its default behaviour that converts
gene names and accession numbers to dates and scientific numbers.
</ol>
<hr>

<h2 id="content">See our YouTube clip</h2>

<p>
 <iframe width="640" height="360" src="https://www.youtube.com/embed/qlWVAHrM4NE" frameborder="0" allowfullscreen></iframe>
</p>
<hr>
<h2 id="content">Get more help</h2>

<p>Contact us by email (mark.ziemann[at]gmail.com).
</p>
<hr>
<hr>
</body>
</html>
