<!DOCTYPE html>

<html xml:lang="en" lang="en">
<link rel='shortcut icon' type='image/x-icon' href='favicon.ico' />
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Digital Expression Explorer 2
</title>
<style type="text/css">

 body{
  -webkit-text-size-adjust:none;
  margin: 20px auto;
  max-width:760px;
  line-height:1.6;
  font:normal normal normal 100% verdana,sans-serif;
  font-size:22px;
  color:#3a3a3a;
  padding:0 10px
 }

 h1,h2,h3{
  line-height:1.2;
  color:#406fef ;
 }

 aside,p,ul{
  color:#3a3a3a;
 }

 table,th,tr,td{
   line-height:1.2;
   font-family:sans-serif;
   font-size:20px;
   background-color:transparent;
   padding: 2px;
   border: 1px solid black;
   border-collapse: collapse;
  }

ul {
    list-style-type: none;
    margin: 0ex;
    padding: 0;
}

li {
    display: inline;
}

.button {
    background-color: #999999;
    color: white;
    padding: 8px 25px;
    leading: 3;
    line-height: 2em;
    text-decoration: none;
#    font-size: 1.8vmax;
    cursor: pointer;
    -webkit-transition-duration: 0.4s; /* Safari */
    transition-duration: 0.4s;
}

.button:hover {
    background-color: #406fef; /* blue */
    color: white;
}

 input{
  font-size:2vmax;
 }

/* The container */
.container {
    display: block;
    position: relative;
    padding-left: 35px;
    margin-bottom: 12px;
    cursor: pointer;
    font-size: 22px;
    color:#3a3a3a;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

/* Hide the browser's default radio button */
.container input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
}

/* Create a custom radio button */
.checkmark {
    position: absolute;
    top: 0;
    left: 0;
    height: 25px;
    width: 25px;
    background-color: #eee;
    border-radius: 50%;
}

/* On mouse-over, add a grey background color */
.container:hover input ~ .checkmark {
    background-color: #ccc;
}

/* When the radio button is checked, add a blue background */
.container input:checked ~ .checkmark {
    background-color: #406fef;
}

/* Create the indicator (the dot/circle - hidden when not checked) */
.checkmark:after {
    content: "";
    position: absolute;
    display: none;
}

/* Show the indicator (dot/circle) when checked */
.container input:checked ~ .checkmark:after {
    display: block;
}

/* Style the indicator (dot/circle) */
.container .checkmark:after {
 	top: 9px;
	left: 9px;
	width: 8px;
	height: 8px;
	border-radius: 50%;
	background: white;
}

</style>
</head>

<body>

<ul>
<a href="/index.html" class="button">HOME</a>
<a href="/about.html" class="button">ABOUT</a>
<a href="/help.html" class="button">HELP</a>
<a href="/bulk.html" class="button">BULK DATA</a>
<a href="/genesets.html" class="button">GENE SETS</a>
</ul>

<div id="page">
	<h1 id="content">News@DEE2</h1>

        <h2 id="content">News archive</h2>

<p> <b>14th March 2022</b> - Data has been transferred and dee2 is operational again.
You will notice that we have https enabled.
We confirm that the R package getDEE2 is also functioning properly.
On the other hand, transfering analyses to Degust is not working and we will
be addressing this next.
</p>

<p> <b>9th March 2022</b> - We now have control of the webserver and we are 
transferring data to a new 4TB volume. 
We estimate that DEE2 will be fully operational on Monday 14th March.
</p>

<p> <b>1st  March 2022 </b> - DEE2 was taken down by the cloud provider due to a 
cybersecurity issue. Apparently the server was found to be sending spam emails.
DEE2 team will restore the service once allowed by cloud provider.
</p>

<p> <b>11th May 2021</b> - We have just completed a migration of DEE2 webserver 
from older Intel Xeon to a new AMD Epyc system.
</p>

<p> <b>3rd Oct 2020</b> - We have curated a number of human disease gene signatures that you can use in enrichment analysis.
More information here: <a href="http://dee2.io/genesets/html">http://dee2.io/genesets.html</a>.
</p>

<p> <b>1st May 2020</b> - Due to popular demand we have made available md5 checksums for the 
bulk data dumps. The file name is "checksums.md5".
These will be updated with each addition of new datasets. Link: 
<a href="http://dee2.io/mx/">http://dee2.io/mx/</a>
</p>

<p> <b>12th March 2020</b> at 11pm today we recorded 1000043 SRA run datasets uploaded to DEE2 
servers. We're expecting the 1 millionth DEE2 dataset to become publicly available within the 
next 1-2 weeks.
</p>

<p> <b>9th January 2020</b> - Happy New Year! Here is a new <a href="http://genomespot.blogspot.com/2020/01/dee2-project-update-jan-2020.html">blog post</a> 
describing recent updates and high priority issues for the next few months.

<br>* Data processing - a work in progress
<br>* Getting a new source of metadata - completed
<br>Future areas for improvement
<br>* The problem of studies missing runs
<br>* Increasing queue efficiency
<br>* Making the R package more efficient
<br>* Website metadata search imptovements
<br>* Call for compute resources
</p>

<p> <b>5th November 2019</b> - the webserver has had a few hiccups recently, but all is restored to normal operation.
In the next few weeks we will be trying to complete the rest of the outstanding mouse and human datasets and then
moving to a new source of metadata.
</p>

<p> <b>20th October 2019</b> - We suffered an unscheduled outage 16-18th Oct 2019 due to human error (sorry!) which was compounded by hardware
issues on the Nectar cloud. All seems to be working now.
</p>

<p> <b>19th June 2019</b> - The June 2019 update to DEE2 has new datasets, new features and new challenges.
  Highlights include new integration with Degust and iDEP and >200k new runs added for human and mouse.
  SRAdbV2 is now permanently dead, so DEE2 wil not be adding any new SRA runs until an alternative is
  found. For the complete story, read on 
  <a href="http://genomespot.blogspot.com/2019/06/dee2-project-update-june-2019.html">here</a>.
</p>

<p> <b>8th April 2019</b> - Very proud to have DEE2 published in Gigascience. Thanks to everyone who has contributed, including reviewers who made several important suggestions. Link <a href="https://doi.org/10.1093/gigascience/giz022" taget="_blank">here</a>.
</p>

<p> <b>11th January 2019</b> - In response to feedback we are unveiling several new features:
</p>
<ul style="list-style-type:circle">
  <li>&bull;Data bundles: any project with >200 runs we've collected into bundles which can be downloaded <a href="/bundles">here</a>. </li><br><br>
  <li>&bull;Keyword searches: users can now string together multiple keywords with "and" to restrict the search results. Multiple word search terms also wrk better now.</li><br><br>
  <li>&bull;Metadata: Now when you perform a search with the web UI, the experiment.name from SRA is shown and hopefully this will make it easier to perform downstream analysis. Also
              we have included full and summary metadata in each DEE2 download. The metadata are also available for download from the bulk page.</li><br><br>
  <li>&bull;R API: previously the API code was sourced from GitHub, due to some security concerns we now have a package on github which can be installed with devtools.</li><br><br>
  <li>&bull;You will also notice two new files in each download GeneInfo.tsv and TxInfo.tsv. These contain the gene symbols information as well as the gene and transcript lengths for FPKM calculation.
              Importantly, the TxInfo.tsv contains the tx-->parent gene information so that transcript data can be easily aggregated to the gene level. In the R package, there is a new function called Tx2Gene
              which performs aggregation (sum) with a single command "x<-Tx2Gene(x)". </li><br><br>
  <li>&bull;QC classification: We have just implemented QC thresholds to classify datasets as pass, warn or fail. More infor on this can be found
              <a href="https://github.com/markziemann/dee2/blob/master/qc/qc_metrics.md" >here</a></li><br><br>
  <li>&bull;Kallisto TPM: In addition to being able to caluculate FPKM using the R API, you can get Kallisto TPM from the repository for individual runs. Here is an example:
              <a href="http://dee2.io/data/dmelanogaster/SRR1991949" target="_blank">SRR1991949</a>. The kallisto output file has the "ke.tsv.gz" suffix.</li><br><br>
</ul>

<p> <b>28th November 2018</b> - DEE2 was presented at the ABACBS2018 conference in Melbourne. See the <a href="https://doi.org/10.5281/zenodo.1561839">Poster</a>.
</p>

<p> <b>28th November 2018</b> - A guide to loading in bulk dumps had been added to my <a href="https://genomespot.blogspot.com/2018/11/using-dee2-bulk-data-dumps.html">Blog</a>.
</p>

<p> <b>12th November 2018 </b>- Ever wondered how accurate DEE2 data actially is? Well we have just undertaken a simulation
study and comparison to GEO deposited data to demonstrate the quality of DEE2 data. More details
<a href="https://github.com/markziemann/dee2/tree/master/validation"> here</a>.
</p>

<p> <b>26th October 2018</b> - Recently there have been many projects with &#8805;500 SRA runs, meaning that they have not 
been readily available from the webserver as one file. In order to address this, I've packaged these SRA projects with 
&#8805;200 SRA run numbers into zip files. Only projects with EVERY run processed successfully by DEE2 are included.
These are now available <a href="/bundles">here</a>.
</p>

<p> <b>7th October 2018</b> - QC information is now linked in the search results and available by hovering the
mouse over the QC link. Note that all the datasets are currently classified as "PASS" but this will be
chenging soon. Also we moved to another nectar server with storage so there have been some updates to
webserver cgi scripts. Also the keyword search is now case insensitive.
</p>

<p> <b>22nd September 2018</b> - the bulk data dumps are now available via http. Dat turned out to be too slow and
unreliable for files of this size.
</p>

<p> <b>21st September 2018</b> - Several small bits of news. Firstly a new domain name server that is working much
better than the last one. Still unable to get https working so this will not likely happen while using
Nectar servers. Secondly the domain name sever change broke the docker image so it was modified and rebuilt.
Thirdly the data processing is progressing steadily with the torrent of new datasets that became visible
upon integration with the newest SRAdbV2. Lastly, the R interface hs undergone several improvements  and
should be more robust now. The new documentation has been added
<a href="https://github.com/markziemann/dee2/blob/master/AccessDEEfromR.md" target="_blank"> (link)</a>.
</p>

<p> <b>14th March 2018</b> - search functionality has been restored, so feel free to test it out.
</p>

<p> <b>1st March 2018</b> - Hope you like the new mobile friendly webpage
</p>

<p> <b>31st January 2018</b> - DEE2 data pipeline is finalised and processing is underway. For more information, see the recent <a href="http://genomespot.blogspot.com.au/2018/01/update-on-dee2-project-for-jan-2018.html" target="_blank"> blog post</a>.
</p>

<p> <b>18th October 2017</b> - As of a month ago, the original DEE site has been down and we've been working towards DEE version 2, which will
be hosted by Monash Uni. This app is still under construction, but stay tuned for the official release at the end 
of 2017.
</p>

<p><b>15th December 2016</b> - Our lab is moving to Monash Central Clinical School. Stay tuned for a revamped DEE in 2017 with new data and features. This URL could go dead anytime next year so please finalise your work.
</p>

<p> <b>13th August 2016</b> - Looking for Illumina bodymap2 count data? Try searching ERP000546 for human.
</p>

<p><b>13th August 2016</b> - Bulk data dumps are back online. An issue with the count matrix headers was resolved.
</p>

<p><b>12th August 2016</b> - Due to an issue with the bulk data dumps, these have been taken down until the problem is resolved. We apologize for the inconvenience.
</p>

<p><b>14th July 2016</b> - Our analysis servers are undergoing upgrades and the July database update will be delayed by approx 2 weeks.
</p>

<p><b>19th Jan 2016</b> - More quality control measures are being added to DEE. This will enable users to view basic QC info such as base qualities, 
number of reads, proportion uniquely mapped, etc. before downloading. These data will be provided to users in parallel with the expression count matrices. 
Enhanced QC data are already available for E coli, C elegans and A thaliana, but are in progress for other species.
</p>

<p>8th Jan 2016</b> - Ever searched for datasets and got the error "Too many results found"? We've modified the interface to display all search results when over the >500 datasets limit. This should allow users to select accession numbers of interest.
</p>

<p><b>8th Jan 2016</b> - A bug in the R function has been resolved. If you had trouble with it before, give it another try and let us know if you encounter any issues getting your data into R.
</p>

<p><b>31st Dec 2015</b> - The latest update run is underway and looking to add about 37k human and 17k mouse datasets. We have all available servers on the job :)
</p>

<p><b>23rd Dec 2015</b> - We've performed an upgrade to SRAdb integration to release some further datasets that were previously missing.
Thanks for your support in 2015 and making it such a successful year. Thanks to your fedback, we're looking forward to a bunch of new features in 2016.
Happy holidays, Merry Christmas and a happy/successful year to you all!
</p>

<p><b>20th Oct 2015</b> - Public announcement of DEE see my <a href="http://genomespot.blogspot.com.au/2015/10/introducing-digital-expression-explorer.html" target="_blank"> blog post</a>.
</p>

<p><b>24th Sept 2015</b> - As of yesterday our site is public, but still in the testing phase.
</p>

<p><b>26th Aug 2015</b> - Bulk and query data are now available in .zip format to improve access for windows and mac users.
</p>

<p><b>24th Aug 2015</b> - Updates to the datasets are now made weekly based on SRAdb entries.
</p>

<p><b>5th Aug 2015</b> - Metadata formatting has been streamlined with SRAdb integration.
</p>

<p><b>17th Jul 2015</b> - More data from human, mouse, Arabidopsis, fruitfly and worm have been added.
</p>

<p><b>9th Feb 2015</b> - The website is under construction.
</p>
<hr>
<hr>
</div>

</body>
</html>


